From 189cad7d809a168488b040cf5510e2d7131e0364 Mon Sep 17 00:00:00 2001
From: Stephan Linz <linz@li-pro.net>
Date: Tue, 17 Dec 2013 12:32:00 +0100
Subject: [PATCH] microblaze: generic: fix missing symbols w/o gpio

Without xilinx gpio driver (systems w/o dedicated gpio controler)
the fdtdec library part will missing symbols: gpio_get_value(),
gpio_set_value(), and gpio_request().

Error is:

  lib/libgeneric.o: In function `fdtdec_get_gpio':
  /u-boot-2013.10/lib/fdtdec.c:506: undefined reference to `gpio_get_value'
  lib/libgeneric.o: In function `fdtdec_set_gpio':
  /u-boot-2013.10/lib/fdtdec.c:516: undefined reference to `gpio_set_value'
  lib/libgeneric.o: In function `fdtdec_setup_gpio':
  /u-boot-2013.10/lib/fdtdec.c:528: undefined reference to `gpio_request'

Signed-off-by: Stephan Linz <linz@li-pro.net>
---
 .../xilinx/microblaze-generic/microblaze-generic.c | 24 +++++++++++++++++++++-
 include/configs/microblaze-generic.h               |  5 ++++-
 2 files changed, 27 insertions(+), 2 deletions(-)

diff --git a/board/xilinx/microblaze-generic/microblaze-generic.c b/board/xilinx/microblaze-generic/microblaze-generic.c
index 42a8d0c..88801e1 100644
--- a/board/xilinx/microblaze-generic/microblaze-generic.c
+++ b/board/xilinx/microblaze-generic/microblaze-generic.c
@@ -17,7 +17,29 @@
 #include <asm/asm.h>
 #include <asm/gpio.h>
 
-#ifdef CONFIG_XILINX_GPIO
+#ifndef CONFIG_XILINX_GPIO
+/*
+ * w/o xilinx gpio driver the fdtdec library part will
+ * missing symbols: gpio_get_value(), gpio_set_value(),
+ * and gpio_request()
+ */
+
+int gpio_get_value(unsigned gpio)
+{
+	return 0;
+}
+
+int gpio_set_value(unsigned gpio, int value)
+{
+	return -1;
+}
+
+int gpio_request(unsigned gpio, const char *label)
+{
+	return -1;
+}
+
+#else
 static int reset_pin = -1;
 #endif
 
diff --git a/include/configs/microblaze-generic.h b/include/configs/microblaze-generic.h
index de02141..bf5b7ca 100644
--- a/include/configs/microblaze-generic.h
+++ b/include/configs/microblaze-generic.h
@@ -296,7 +296,10 @@
 #define CONFIG_CMD_IRQ
 #define CONFIG_CMD_MFSL
 #define CONFIG_CMD_ECHO
-#define CONFIG_CMD_GPIO
+
+#if defined(CONFIG_XILINX_GPIO)
+# define CONFIG_CMD_GPIO
+#endif
 
 #define CONFIG_CMD_ROMFS
 
-- 
1.8.3.4

